### PR TITLE
Introduce ClassMetadataFactoryInterface to allow switching implementations

### LINKED ISSUE #2551: Cannot override the class metadata factory via configuration since 2.5.0
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.5.0

#### Summary

Since version 2.5.0, the class metadata factory cannot be overridden.

#### Previous behavior

The document manager has been instantiated taking the class name from the configuration on document manager construction

#### Current behavior

Since https://github.com/doctrine/mongodb-odm/commit/5987555a4a2f19125089f0061a978f1d8b9e818f the override is broken as the document manager is wired to the final class `Doctrine\ODM\MongoDB\Mapping\ClassMetadataFactory`.
To allow overriding via configuration, it should be wired against `Doctrine\Persistence\Mapping\ClassMetadataFactory` interface.

#### How to reproduce

```php
$config = new Configuration();
$config->setClassMetadataFactoryName(FakeMetadataFactory::class);
DocumentManager::create(config: $config);
```

### PR DESCRIPTION
**Pull request by malarzm**
|      Q       |   A
|------------- | -----------
| Type         | bugfix
| BC Break     | fixing one
| Fixed issues | fixes #2551

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/DocumentManager.php [hunk 1]
```diff
@@ -155,10 +157,13 @@ protected function __construct(?Client $client = null, ?Configuration $config =
             ],
         );
 
+        $this->classNameResolver = new CachingClassNameResolver(new ProxyManagerClassNameResolver($this->config));
+
         $metadataFactoryClassName = $this->config->getClassMetadataFactoryName();
         $this->metadataFactory    = new $metadataFactoryClassName();
         $this->metadataFactory->setDocumentManager($this);
         $this->metadataFactory->setConfiguration($this->config);
+        $this->metadataFactory->setProxyClassNameResolver($this->classNameResolver);
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/DocumentManager.php:166]: Rearranging so initialization is done in one place

File: lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Validation.php [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 use Attribute;
 use Doctrine\Common\Annotations\Annotation\NamedArgumentConstructor;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Validation.php]: Unrelated, but phpcbf fixed this automatically

**Review by alcaeus — COMMENTED** [review]
Static analysis isn't happy, but LGTM in general.

**Review by IonBazan — APPROVED** [review]

**Comment by malarzm** [comment]
Rebasing as #2553 brought back changes for SA/CS tools to 2.5.x branch. Will work on SA happiness and merge this when CI is green :)

**Pull request merged**